text/3334-niche.md

```diff
@@ -0,0 +1,480 @@
+  multiple discontiguous niches, the compiler need not take all of them into
+  account.)
+- A raw pointer. (This allows user-defined types to store a properly typed
+  pointer while using known-invalid pointer values as niches.)
+- A fieldless enum with a `repr` of a primitive integer type.
+
+Declaring a niche on a struct whose field type does not meet these restrictions
+results in an error.
+
+Declaring a niche on any item other than a struct declaration results in an
+error.
+
+Declaring a niche on a struct containing more or less than one field results in
+an error.
+
+Declaring multiple `niche` attributes on a single item, or multiple key-value
+pairs within a single `niche` attribute, results in an error.
+
+The `niche` attribute may appear inside `cfg_attr`. The net effect after
+evaluating all configuration must be to apply either zero or one `niche`
+attribute to the type.
+
+Declaring a niche on a struct that has any generic parameters results in an
+error.
+
+Declaring a range niche with an empty range (e.g. `0..0`) results in a
+warn-by-default lint. As with many lints, this lint should be automatically
+suppressed for code expanded from a macro.
+
+Declaring a range niche with an invalid range (e.g. `5..0`) results in an
+error.
+
+Declaring a niche using a negative value or a negative range endpoint results
+in an error. The representation of negative values depends on the size of the
+type, and the compiler may not have that information at the time it handles
+attributes such as `niche`. The text of the error should suggest the
+appropriate two's-complement unsigned equivalent to use. The compiler may
+support this in the future.
+
+Declaring a range niche with an open start (`..3`) results in an error, for
+forwards-compatibility with support for negative values.
+
+Declaring a niche using a non-literal value (e.g. `usize::MAX`) results in an
+error. Constants can use compile-time evaluation, and compile-time evaluation
+does not occur early enough for attributes such as niche declarations.
+
+If a type `T` contains multiple niche values (e.g. `#[niche(range = 8..16)]`),
+the compiler does not guarantee any particular usage of those niche values in
+the representation of types containing `T`. In particular, the compiler does
+not commit to making use of all the niche values, even if it otherwise could
+have.
+
+However, multiple instances of the same identical type (e.g. `Option<T>` and
+`Option<T>`) will use an identical representation (whether the type contains a
+niche or not). This permits a round-trip between such a value and a byte
+representation.
+
+Adding a niche to a type does not change the storage size of the type, even if
+the niche might otherwise allow storing fewer bytes. The type still allows
+obtaining mutable references to the field, which requires storing valid values
+using the same representation as those values would have had without the niche.
```

Comment:
I would strengthen this to be basically verbatim as the [layout section of to docs as `MaybeUninit`](https://doc.rust-lang.org/std/mem/union.MaybeUninit.html#layout-1), basically:

1. The struct has the same size, alignment, and ABI as the contained type.
2. This invariant is intentionally broken when nesting types, since putting the struct inside an `Option` is not the same layout as the `Option` with the original type. (`Option` isn't the only case here, it's just a useful example.)